### ISSUE TITLE
WW3 not producing same answers with different number of pets in UFS-weather-model

### ISSUE THREAD

**Issue by JessicaMeixner-NOAA** [NOAA-EMC/WW3#317]
When changing the number of pets for the wave model in the UFS-weather-model the restart file and other files are not reproducing themselves.  Need to determine the cause and fix this.

**Comment by aliabdolali** [comment]
@JessicaMeixner-NOAA The simple standalone test to check the number of tasks and thread reproducibility is addressed here, right?
 https://github.com/aliabdolali/WW3/blob/fb_ufs_regtests/regtests/bin/matrix.base#L2090-L2099

**Comment by JessicaMeixner-NOAA** [comment]
This is a close replica of the ufs-weather-model for testing this.   The differences are likely in the time stepping and the fact that the ufs-weather-model tests separate the binary output by hour. I also don't know if it's the same frequency of restart or other output, I forgot to check that yesterday.   Is that test reproducing itself for different number of mpi tasks?

**Comment by JessicaMeixner-NOAA** [comment]
@MatthewMasarik-NOAA @aliabdolali and I have been working on this issue.  Some updates as of now: 
* Added tests in a branch of ufs-weather-model to test (note the C96 test passes the mpi test but the C384 bmark does not): https://github.com/jessicameixner-noaa/ufs-weather-model/tree/feature/addwavetests 
* In working with the WW3 regression tests, ww3_ufs1.1 we have found that the reproducibility issue is at 12z in W3UICE, but interestingly if you make an ice field that updates more frequently and do not go through 12z it does reproduce.  
* When running the ww3_ufs1.1 test with debug flags, there is an error.

**Comment by aliabdolali** [comment]
I created an equivalent setup for ww3_ufs1.1 for 5 degree grid with ww3_shel instead of ww3_multi
In ww3_shell. we have the following options:
```
$ Define output data ------------------------------------------------- $
$
$ Define output server mode. This is used only in the parallel version
$ of the model. To keep the input file consistent, it is always needed.
$ IOSTYP = 1 is generally recommended. IOSTYP > 2 may be more efficient
$ for massively parallel computations. Only IOSTYP = 0 requires a true
$ parallel file system like GPFS.
$
$    IOSTYP = 0 : No data server processes, direct access output from
$                 each process (requirese true parallel file system).
$             1 : No data server process. All output for each type
$                 performed by process that performes computations too.
$             2 : Last process is reserved for all output, and does no
$                 computing.
$             3 : Multiple dedicated output processes.
$
```
> Type 1, 2, 3 between 12 and 24 cores behaves similar to ww3_multi (first few restart files are identical, then different)
Type 0 is always different. 

> I removed ice from forcing and everything was identical.

**Comment by MatthewMasarik-NOAA** [comment]
In addition to @JessicaMeixner-NOAA comments above regarding W3UICE, we've found another area that will affect the outcome of the ww3_ufs1.1 regtest.  Currently, we're looking at the block of code in W3IORS that collects the action density (`VA`/`VAAUX`), copies it to buffer `WRITEBUFF`, then writes it to  restart file.  If you comment out this write to file, the restarts are all identical.

**Comment by MatthewMasarik-NOAA** [comment]
Comparing the VA/VAAUX values for different MPI counts from the code block in W3IORS mentioned above, there is one pattern that sticks out.  There are consistently four global sea points (ISEA) which have differing spectral element values.  These points and global 2D coordinates (note: for _glo_5deg_) are:
* __ISEA__:  \[__IY__, __IX__\]
* 34: \[4, 37\]
* 37: \[4, 40\]
* 84: \[5, 18\]
* 88: \[5, 22\]

There's no obvious pattern in the number or position of spectral elements that differ at these four sea points for different MPI counts.  However, comparing across a range of MPI counts, these four points and these only, are the ones having different values.

**Comment by MatthewMasarik-NOAA** [comment]
The sea points above were found to all be points that were de-activated in the W3UICE update.  Running 1 deg for multiple MPI counts found the same behavior.  The re-activation points have no problems, so we've narrowed the issue to be related to de-activation of sea points in the ice concentration update.

**Comment by aliabdolali** [comment]
The problem persists with different versions of intel/impi compiler on Hera.
The problem persists with gnu/mpich.

**Comment by JessicaMeixner-NOAA** [comment]
@MatthewMasarik-NOAA @aliabdolali and I have traced this back to likely being an issue that in w3scat we do not update the values of VA for points that have MAPSTA=-1.  While you would think that this should be 0 if you recently have an ice covered point, after w3scat we found this to not be the case on a point that was just deactivated (IC0 MAPSTA=1 going to -1).  This seems to correct other issues reported for MPI issues as well and we got the reproducibility desired for the MPI test with ufs-weather-model; however, we had other answers change that we were not expecting.  These were bundled with some clean-up of W3UICE which we discovered cause inexplicable changes to mod_defs in a few cases.   Testing on just changing w3scat is ongoing and we hope to submit a PR with this update soon, assuming the isolated w3scat change produces the desired reproducible results.
